src/Nethermind/Nethermind.Core/Threading/ParallelUnbalancedWork.cs

```diff
@@ -0,0 +1,332 @@
+    /// <param name="parallelOptions">An object that configures the behavior of this operation.</param>
+    /// <param name="state">The initial state of the thread-local data.</param>
+    /// <param name="action">The delegate that is invoked once per iteration.</param>
+    public static void For<TLocal>(
+        int fromInclusive,
+        int toExclusive,
+        ParallelOptions parallelOptions,
+        TLocal state,
+        Func<int, TLocal, TLocal> action)
+        => InitProcessor<TLocal>.For(fromInclusive, toExclusive, parallelOptions, null, state, action);
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ParallelUnbalancedWork"/> class.
+    /// </summary>
+    /// <param name="data">The shared data for the parallel work.</param>
+    private ParallelUnbalancedWork(Data data)
+    {
+        _data = data;
+    }
+
+    /// <summary>
+    /// Executes the parallel work item.
+    /// </summary>
+    public void Execute()
+    {
+        int i = _data.Index.GetNext();
+        while (i < _data.ToExclusive)
+        {
+            _data.Action(i);
+            // Get the next index
+            i = _data.Index.GetNext();
+        }
+
+        // Signal that this thread has completed its work
+        _data.MarkThreadCompleted();
+    }
+
+    /// <summary>
+    /// Provides a thread-safe counter for sharing indices among threads.
+    /// </summary>
+    private class SharedCounter(int fromInclusive)
+    {
+        private PaddedValue _index = new(fromInclusive);
+
+        /// <summary>
+        /// Gets the next index in a thread-safe manner.
+        /// </summary>
+        /// <returns>The next index.</returns>
+        public int GetNext() => Interlocked.Increment(ref _index.Value) - 1;
+
+        [StructLayout(LayoutKind.Explicit, Size = 128)]
+        private struct PaddedValue(int value)
+        {
+            [FieldOffset(64)]
+            public int Value = value;
+        }
+    }
+
+    /// <summary>
+    /// Represents the base data shared among threads during parallel execution.
+    /// </summary>
+    private class BaseData(int threads, int fromInclusive, int toExclusive)
+    {
+        /// <summary>
+        /// Gets the shared counter for indices.
+        /// </summary>
+        public SharedCounter Index { get; } = new SharedCounter(fromInclusive);
+        public SemaphoreSlim Event { get; } = new(initialCount: 0);
+        private int _activeThreads = threads;
+
+        /// <summary>
+        /// Gets the exclusive upper bound of the range.
+        /// </summary>
+        public int ToExclusive => toExclusive;
+
+        /// <summary>
+        /// Gets the number of active threads.
+        /// </summary>
+        public int ActiveThreads => Volatile.Read(ref _activeThreads);
+
+        /// <summary>
+        /// Marks a thread as completed.
+        /// </summary>
+        /// <returns>The number of remaining active threads.</returns>
+        public int MarkThreadCompleted()
```

Comment:
minor: in theory this could be misused by calling `MarkThreadCompleted` first, but as this is private class we can ignore it.